src/main/scala/common/scratchpad_adapter.scala

```diff
@@ -0,0 +1,115 @@
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util._
+
+class SodorScratchpadAdapter(implicit p: Parameters, implicit val sodorConf: SodorConfiguration) extends Module {
+  // Parameter traits
+  val coreParams = {
+    class C(implicit val p: Parameters) extends HasCoreParameters
+    new C
+  }
+  // Extract tileParams from HasNonDiplomaticTileParameters, which is the base trait of HasCoreParameters above
+  val tileParams = coreParams.tileParams
+  // Sodor config class
+
+  // Sodor constants
+  val sodorConst = {
+    class S extends MemoryOpConstants
+    new S
+  }
+
+  val io = IO(new Bundle() {
+    val slavePort = Flipped(new HellaCacheIO())
+    val memPort = new MemPortIo(data_width = coreParams.coreDataBits)
+  })
+
+  // ===================
+  // Slave port signals
+  val slave_req_ready = io.slavePort.req.ready
+  val s2_slave_resp_valid = io.slavePort.resp.valid
+  val slave_req_valid = io.slavePort.req.valid
+
+  val slave_cmd = io.slavePort.req.bits.cmd
+  val slave_req_addr = io.slavePort.req.bits.addr(log2Ceil(tileParams.dcache.get.dataScratchpadBytes), 0)
+  val slave_req_size = io.slavePort.req.bits.size
+  val slave_req_signed = io.slavePort.req.bits.signed
+
+  // All request are delayed for one cycle to avoid being killed
+  val s1_slave_write_kill = io.slavePort.s1_kill
+  val s1_slave_write_data = io.slavePort.s1_data.data
+  val s1_slave_write_mask = io.slavePort.s1_data.mask
+
+  val s1_slave_req_valid = RegNext(slave_req_valid)
+  val s1_slave_cmd = RegNext(slave_cmd)
+  val s1_slave_req_addr = RegNext(slave_req_addr)
+  val s1_slave_req_size = RegNext(slave_req_size)
+  val s1_slave_req_signed = RegNext(slave_req_signed)
+
+  // Note that ScratchpadSlavePort requires 2-cycle delay, or it won't even send the response
+  val s2_slave_read_data = io.slavePort.resp.bits.data_raw
+  val s2_slave_read_mask = io.slavePort.resp.bits.mask
+
+  val s2_nack = io.slavePort.s2_nack
+
+  // Tie anything not defined below to DontCare
+  io.slavePort := DontCare
+
+  // ===================
+  // HellaCacheIO to MemPortIo logic
+  // Connect valid & ready bits
+  slave_req_ready := io.memPort.req.ready
+  io.memPort.req.valid := s1_slave_req_valid & (s1_slave_cmd === M_XRD || !s1_slave_write_kill)
+  s2_slave_resp_valid := RegNext(io.memPort.resp.valid)
+
+  // Connect read info
+  s2_slave_read_mask := RegNext(new StoreGen(s1_slave_req_size, s1_slave_req_addr, 0.U, coreParams.coreDataBytes).mask)
+  s2_slave_read_data := RegNext(io.memPort.resp.bits.data)
+
+  // Connect write info
+  io.memPort.req.bits.addr := s1_slave_req_addr
+  io.memPort.req.bits.data := s1_slave_write_data
+
+  // Other connections
+  s2_nack := false.B
+  io.memPort.req.bits.fcn := Mux(s1_slave_cmd === M_XRD, sodorConst.M_XRD, sodorConst.M_XWR)
+  // Since we don't have dword here (the bus only has 32 bits), s1_slave_req_size <= 2.
+  // The expression below convert TileLink size and signedness to Sodor type.
```

Comment:
Lets add a `require` somewhere checking that the width of the tilelink port is 32.

Reply:
SodorScratchpadAdapter does not see TileLink IO directly, so there is no way to check the width. However, I added an `assert` here to check the slave port request size, and it should report an error shall an oversize request come in.

Reply:
You can add this check in the Sodor Tile, where the TileLink IO is. We want to prevent system misconfiguration

Reply:
Never mind, I figured out how to add the check here.